x-pack/legacy/plugins/lens/public/indexpattern_plugin/datapanel.tsx

```diff
@@ -50,11 +50,55 @@ const fieldTypeNames: Record<DataType, string> = {
   date: i18n.translate('xpack.lens.datatypes.date', { defaultMessage: 'date' }),
 };
 
-export function IndexPatternDataPanel(props: DatasourceDataPanelProps<IndexPatternPrivateState>) {
+export function IndexPatternDataPanel({
+  setState,
+  state,
+  dragDropContext,
+  state: { indexPatterns, currentIndexPatternId },
+}: DatasourceDataPanelProps<IndexPatternPrivateState>) {
+  const [showIndexPatternSwitcher, setShowIndexPatternSwitcher] = useState(false);
+
+  const onChangeIndexPattern = useCallback(
+    (newIndexPattern: string) => {
+      setState({
+        ...state,
+        currentIndexPatternId: newIndexPattern,
+      });
+    },
+    [state, setState]
+  );
+
+  return (
+    <MemoizedDataPanel
+      showIndexPatternSwitcher={showIndexPatternSwitcher}
+      setShowIndexPatternSwitcher={setShowIndexPatternSwitcher}
+      currentIndexPatternId={currentIndexPatternId}
+      indexPatterns={indexPatterns}
+      dragDropContext={dragDropContext}
+      // only pass in the state change callback if it's actually needed to avoid re-renders
+      onChangeIndexPattern={showIndexPatternSwitcher ? onChangeIndexPattern : undefined}
```

Comment:
This is fine, but I wonder if we should move from a setState kind of approach to a dispatch kind of approach, for this reason. dispatch-helpers can be easily memoized as they don't need to close over the state object.